### PR TITLE
Bugfix for DOEDriver full factorial array inputs

### PR DESCRIPTION
**Pull request by onodip**
This PR fixes a bug of the `FullFactorialGenerator` in the `DOEDriver`

**Bug:**
Wrong factoring, if there exists at least one array design variable with array size larger than one, and it is not the last one. The array test of this class was testing with only one design variable, therefore it was not caught before.

**Fix:**
Corrected a wrong indexing in `doe_generators.py`. Location in the comments
Added a new test.

I also made some other small changes for performance and readability

The only test so far, where it was used with multiple array inputs, was in `test_csv_array()`. Here the expected result was not really a full factorial, since it included non-unique cases (each repeated twice), as the figure below shows. 

image

### CODE REVIEW

File: openmdao/drivers/tests/test_doe_driver.py [hunk 1]
```diff
@@ -337,36 +318,36 @@ def test_csv_array(self):
         prob.run_driver()
         prob.cleanup()
 
-        expected = {
```
Comment (onodip) [inline openmdao/drivers/tests/test_doe_driver.py]: This is the expected result shown in the figure. 

Which is equivalent to:
`pyDOE2.fullfact(levels=[2, 2, 2])
`

Which correctly results in:
```
array([[0., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [1., 1., 0., 0.],
       [0., 0., 1., 0.],
       [1., 0., 1., 0.],
       [0., 1., 1., 0.],
       [1., 1., 1., 0.],
       [0., 0., 0., 1.],
       [1., 0., 0., 1.],
       [0., 1., 0., 1.],
       [1., 1., 0., 1.],
       [0., 0., 1., 1.],
       [1., 0., 1., 1.],
       [0., 1., 1., 1.],
       [1., 1., 1., 1.]])
```

File: openmdao/drivers/doe_generators.py [hunk 2]
```diff
@@ -347,26 +349,24 @@ def __call__(self, design_vars, model=None):
...
         for idxs in doe.astype('int'):
             retval = []
-            var = row = 0
-            for name, meta in iteritems(design_vars):
+            row = 0
+            for var, (name, meta) in enumerate(iteritems(design_vars)):
                 size = meta['size']
                 val = np.empty(size)
                 for k in range(size):
-                    idx = idxs[var + k]
```
Comment (onodip) [inline openmdao/drivers/doe_generators.py]: This is the actual bug. `var` was the index of the design variable, but array DVs take multiple rows in `idxs`

**Comment by onodip** [comment]
No idea, how the error in the tests is related to the changes.

**Comment by swryan** [comment]
> No idea, how the error in the tests is related to the changes.

see: https://github.com/onodip/OpenMDAO/pull/3

**Comment by onodip** [comment]
Thanks for the help @swryan

**Pull request merged**